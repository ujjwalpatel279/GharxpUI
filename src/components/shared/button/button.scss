// ========================================================================
// COMPONENTS/SHARED/BUTTON
// ========================================================================

@import '~sassHelpers';

.c-btn {
  position: relative;
  display: inline-flex;
  align-items: center;
  height: spacer(10);
  padding-left: spacer(8);
  padding-right: spacer(8);
  justify-content: center;
  border-radius: $border-radius;
  transition: all 0.1s ease-in;

  &:hover,
  &:focus {
    opacity: 0.9;
    transform: scale(1.02);
  }

  &:focus {
    outline: 1px dashed color-brand(primary);
    outline-offset: 2px;
  }

  &:disabled {
    opacity: 0.5;
    cursor: default;

    &:hover,
    &:focus {
      opacity: 0.5;
      transform: none;
    }
  }

  svg {
    font-size: 12px;
  }

  &--primary {
    color: white;
    background-color: color-brand(primary);

    svg {
      fill: white;
    }
  }

  &--secondary {
    color: color-brand(primary);
    background-color: color-brand(primary-light-2);

    svg {
      fill: color-brand(primary);
    }
  }

  &--outline {
    color: color-brand(primary);
    border: 1px solid color(grey3);

    svg {
      fill: color-brand(primary);
    }
  }

  &--navy {
    background-color: color(navy1);
    color: white;

    svg {
      fill: white;
    }
  }

  &--bare {
    padding: spacer(2) 0;
  }

  &--close {
    width: 30px;
    height: 30px;
    padding: 2px;
    background: color(grey1);
    border-radius: 50%;

    svg {
      fill: white;
    }
  }

  &--small {
    height: spacer(8);
    padding-left: spacer(4);
    padding-right: spacer(4);
  }

  &--icon-left svg {
    margin-right: spacer(2);
  }

  &--icon-right svg {
    margin-left: spacer(2);
  }

  &--text-hidden span {
    @include hidden-visually;
  }

  &__larger_screens_hidden {
    @include bp(10) {
      display: none;
    }
  }
}
