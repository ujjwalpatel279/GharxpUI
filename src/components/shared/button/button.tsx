import React, { FunctionComponent, ReactElement, MouseEvent } from 'react';
import { Icon, IconNames } from '../icon/icon';

import './button.scss';

export interface ButtonProps {
  type: 'button' | 'submit' | 'reset';
  variant: 'primary' | 'secondary' | 'outline' | 'navy' | 'bare';
  size?: 'small' | 'large';
  width?: 'full';
  dataValue?: string | number;
  disabled?: boolean;
  childClasses?: string;
  icon?: IconNames;
  iconPosition?: 'left' | 'right';
  textHidden?: boolean;
  onClick?: (e: MouseEvent<HTMLButtonElement>) => void;
  hiddenInLargerScreens?: boolean;
  visibility?: boolean;
}

export const Button: FunctionComponent<ButtonProps> = ({
  type,
  variant,
  size,
  width,
  dataValue,
  disabled,
  childClasses,
  icon,
  iconPosition,
  textHidden,
  onClick,
  children,
  hiddenInLargerScreens,
}): ReactElement => {
  let btnClasses = `c-btn c-btn--${variant}`;

  if (size) {
    btnClasses += ` c-btn--${size}`;
  }

  if (width === 'full') {
    btnClasses += ` u-width-100`;
  }

  if (disabled) {
    btnClasses += ` disabled`;
  }

  if (childClasses) {
    btnClasses += ` ${childClasses}`;
  }

  if (iconPosition === 'left') {
    btnClasses += ' c-btn--icon-left';
  }

  if (iconPosition === 'right') {
    btnClasses += ' c-btn--icon-right';
  }

  if (textHidden) {
    btnClasses += ' c-btn--text-hidden';
  }

  if (hiddenInLargerScreens) {
    btnClasses += ` c-btn__larger_screens_hidden`;
  }

  return (
    <button
      type={type}
      className={btnClasses}
      disabled={disabled}
      {...(dataValue && { 'data-value': dataValue })}
      onClick={onClick}
    >
      {icon && (iconPosition === 'left' || textHidden) && <Icon name={icon} />}
      <span>{children}</span>
      {icon && iconPosition === 'right' && <Icon name={icon} />}
    </button>
  );
};
