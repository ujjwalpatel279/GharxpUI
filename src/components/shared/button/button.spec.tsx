import React from 'react';
import { render, cleanup } from '@testing-library/react';

import { Button } from './button';

describe('Button Component', () => {
  afterEach(cleanup);

  it('renders the primary button', () => {
    const { container } = render(
      <Button type="button" variant="primary">
        Primary Button
      </Button>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders the primary button small', () => {
    const { container } = render(
      <Button type="button" variant="primary" size="small">
        Primary Button Small
      </Button>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders the primary button large', () => {
    const { container } = render(
      <Button type="button" variant="primary" size="large">
        Primary Button Large
      </Button>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders the secondary button', () => {
    const { container } = render(
      <Button type="button" variant="secondary">
        Secondary Button
      </Button>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders the secondary button Small', () => {
    const { container } = render(
      <Button type="button" variant="secondary" size="small">
        Secondary Button Small
      </Button>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders the secondary button large', () => {
    const { container } = render(
      <Button type="button" variant="secondary" size="large">
        Secondary Button Large
      </Button>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders the primary button navy', () => {
    const { container } = render(
      <Button type="button" variant="navy">
        Button Navy
      </Button>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders a disabled button', () => {
    const { container } = render(
      <Button type="button" variant="primary" disabled>
        Button Disabled
      </Button>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders a full width button', () => {
    const { container } = render(
      <Button type="button" variant="primary" width="full">
        Button Full Width
      </Button>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
