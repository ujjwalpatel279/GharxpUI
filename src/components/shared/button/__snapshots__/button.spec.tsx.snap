// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component renders a disabled button 1`] = `
<button
  class="c-btn c-btn--primary disabled"
  disabled=""
  type="button"
>
  <span>
    Button Disabled
  </span>
</button>
`;

exports[`Button Component renders a full width button 1`] = `
<button
  class="c-btn c-btn--primary u-width-100"
  type="button"
>
  <span>
    Button Full Width
  </span>
</button>
`;

exports[`Button Component renders the primary button 1`] = `
<button
  class="c-btn c-btn--primary"
  type="button"
>
  <span>
    Primary Button
  </span>
</button>
`;

exports[`Button Component renders the primary button large 1`] = `
<button
  class="c-btn c-btn--primary c-btn--large"
  type="button"
>
  <span>
    Primary Button Large
  </span>
</button>
`;

exports[`Button Component renders the primary button navy 1`] = `
<button
  class="c-btn c-btn--navy"
  type="button"
>
  <span>
    Button Navy
  </span>
</button>
`;

exports[`Button Component renders the primary button small 1`] = `
<button
  class="c-btn c-btn--primary c-btn--small"
  type="button"
>
  <span>
    Primary Button Small
  </span>
</button>
`;

exports[`Button Component renders the secondary button 1`] = `
<button
  class="c-btn c-btn--secondary"
  type="button"
>
  <span>
    Secondary Button
  </span>
</button>
`;

exports[`Button Component renders the secondary button Small 1`] = `
<button
  class="c-btn c-btn--secondary c-btn--small"
  type="button"
>
  <span>
    Secondary Button Small
  </span>
</button>
`;

exports[`Button Component renders the secondary button large 1`] = `
<button
  class="c-btn c-btn--secondary c-btn--large"
  type="button"
>
  <span>
    Secondary Button Large
  </span>
</button>
`;
